SwiftCompilerSources/Sources/Optimizer/Utilities/AccessUtils.swift

```diff
@@ -0,0 +1,421 @@
+  var isUniquelyIdentified: Bool {
+    switch baseKind {
+    case .box:
+      // The operand `%op` in `%baseAddress = project_box %op` can
+      // be `alloc_box` or an argument. Only if it's a fresh allocation it is
+      // uniquelyIdentified, otherwise it is aliasable, as all the other references.
+      return baseAddress is AllocBoxInst
+    case .stack, .global:
+      return true
+    case .argument:
+      // An argument address that is non-aliasable
+      return (baseAddress as! FunctionArgument).isExclusiveIndirectParameter
+    case .class, .tail, .yield, .pointer:
+      // References (.class and .tail) may alias, and so do pointers and
+      // yield results
+      return false
+    }
+  }
+
+  /// Returns `true` if the two access bases do not alias
+  func isDistinct(from other: AccessBase) -> Bool {
+    switch (baseAddress, other.baseAddress) {
+    case is (AllocStackInst, AllocStackInst),
+         is (AllocBoxInst, AllocBoxInst):
+      return baseAddress != other.baseAddress
+    case let (this as GlobalAddrInst, that as GlobalAddrInst):
+      return this.global != that.global
+    case let (this as FunctionArgument, that as FunctionArgument):
+      return (this.isExclusiveIndirectParameter || that.isExclusiveIndirectParameter) && this != that
+    case let (this as RefElementAddrInst, that as RefElementAddrInst):
+      return (this.fieldIndex != that.fieldIndex)
+    default:
+      if isUniquelyIdentified && other.isUniquelyIdentified && baseKind != other.baseKind { return true }
+      // property: `isUniquelyIdentified` XOR `isObject`
+      if isUniquelyIdentified && other.baseKind.isObject { return true }
+      if baseKind.isObject && other.isUniquelyIdentified { return true }
+      return false
+    }
+  }
+}
+
+/// An `AccessPath` is a pair of a `base: AccessBase` and a `projectionPath: Path`
+/// which denotes the offset of the access from the base in terms of projections.
+struct AccessPath : CustomStringConvertible {
+  var base: AccessBase
+
+  /// address projections only
+  var projectionPath: SmallProjectionPath
+
+  var description: String {
+    "\(projectionPath): \(base)"
+  }
+
+  func isDistinct(from other: AccessPath) -> Bool {
+    return base.isDistinct(from: other.base) ||
+      (base.baseAddress == other.base.baseAddress && projectionPath != other.projectionPath)
+  }
+}
+
+/// An `AccessStoragePath` is the reference (or a value for which the
+/// reference is a stored property) an address originates from.
+struct AccessStoragePath {
+  let storage: Value
+
+  /// Only valid paths are: `"<sequence of value projections>.<one reference projection>.<sequence of address projections>"`
+  let path: SmallProjectionPath
+}
+
+typealias AccessStorage = (maybeRoot: AccessStoragePath?, base: AccessBase)
+
+enum EnclosingScope {
+  case scope(BeginAccessInst)
+  case base(AccessBase)
+}
+
+
+func canBeOperandOfIndexAddr(_ value: Value) -> Bool {
```

Comment:
It's not clear if this is a strict SIL invariant or if this just defines the kind of patterns which we accept here.
I would assume, that it's the second (for now - we might verify that in future). In this case, please make it a `private` function. Then it's clear it's only used in the access utils.